text/0000-offline-buildpackages.md

```diff
@@ -0,0 +1,236 @@
+- Explaining the feature largely in terms of examples.
+
+If we want to support offline builds, with the current lifecycle & buildpackage spec we are forced to throw all dependencies inside of a buildpack archive. So an offline `ruby` buildpack would have a structure similar to the following:
+
+```
+cnb
+├─ buildpacks
+│   └─ ruby-buildpack
+│       └─ 1.2.3
+│           ├── bin
+│           │   ├── build
+│           │   └── detect
+│           ├── dependencies
+│           │   ├── ruby-v2.7.1
+│           │   └── ruby-v2.6.8
+│           └── buildpack.toml
+└── lifecycle
+```
+There are problems with the above:
+
+- Buildpacks packaged with dependencies & packaged without dependencies have different `sha256` values and therefore must have different buildpack `id` fields to avoid registry collision issues.
+
+We propose a `lifecycle` mechanism for exposing buildpack dependencies, and an extension to buildpackages that moves these dependencies out of buildpack archives.
+
+1) Buildpack dependencies should be made available during the `build` phase in a file tree rooted at `/cnb/deps`
+2) Buildpacks should be able to access dependencies they were packaged with via the `CNB_DEPS` environment variable.
+3) The [`package.toml`](https://buildpacks.io/docs/reference/package-config/) file should have additional metadata for each dependencies to specify
+    - the endpoint a dependency is exposed at via the lifecycle 
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Answering the following should outline the behavior of this feature.
+
+### How will dependencies appear on the filesystem during the `build` phase?
+
+Dependencies will be listed underneath `/cnb/deps/` with a schema that has the following properties:
+
+- dependency placement does not depend on the dependencies bits (no hashes, SHA256, ...).
+- dependency placement is calculable by the lifecycle (needed to provide through `CNB_DEPS`).
+- dependency placement is specific enough to reduce inter-buildpack collisions.
+- dependency placement provides compartmentalization between buildpacks, (`CNB_DEPS` only provides paths to dependencies a buildpack should have access to).
+
+```
+cnb
+├─ buildpacks
+│   └─ ruby-buildpack
+│       └─ 1.2.3
+│           ├── bin
+│           │   ├── build
+│           │   └── detect
+│           └── buildpack.toml
+├─ deps
+│   ├── io.paketo.mri
+│   │    ├── v2.7.1.toml
+│   │    ├── v2.7.1
+│   │    ├── v2.6.8.toml
+│   │    └── v2.6.8
+│   ├── io.paketo.other-dep
+│   │    ├── v2.2.2.toml
+│   │    └── v2.2.2
+│   └── <id>
+│        ├── v1.2.3.toml
+│        ├── v1.2.3
+│        ├── <version>
+│        └── <version>.toml
+└── lifecycle
+```
+### How will paths to dependencies be exposed to a buildpack?
+
+During the `build` phase of the `paketo-buildpacks/mri` buildpack, the `CNB_DEPS` environment variable will provide a path pointing to a `packaged.toml` file, this will contain information on all all dependencies packaged with a buildpack.
```

Comment:
```suggestion
During the `build` phase of the `paketo-buildpacks/mri` buildpack, the `CNB_DEPS` environment variable will provide a path pointing to a `packaged.toml` file, this will contain information on all dependencies packaged with a buildpack.
```